<div id="container_rightContent">

% if (@err) {
<div style="display:none" id="error_list">
%	foreach (@err) {
<div><% $_ %></div>
%	}
</div>
% }

<form id="forma1" enctype="multipart/form-data" method="post">
<div id="conRedlineLeft1"> 
	<div id="conProject_newLeft"> 

		<div class="conNewPro_label">
			<div class="conNewPro_title">Select Organism type *</div>
			<div class="conNewPro_label1">
				<span class="conNewPro_smallContainerx">
				<input class="conRadiobox_align" type="radio" name="otype" id="otypea" value="animals" <% $otype eq "animals" ? "checked=\"checked\"" : ""%>/><label for="otypea"> Animal</label></span>
				<span class="conNewPro_smallContainerx">
				<input class="conRadiobox_align" type="radio" name="otype" id="otypep" value="plants" <% $otype eq "plants" ? "checked=\"checked\"" : ""%>/><label for="otypep"> Plant</label></span>
<!--				<span class="conNewPro_smallContainerx">
				<input class="conRadiobox_align" type="radio" name="otype" id="otypef" value="fungi" <% $otype eq "fungi" ? "checked=\"checked\"" : ""%>/><label for="otypef"> Fungi</label></span>
-->
			</div>
			<div style="clear:both;height:5px;">&nbsp;</div>
			<div style="height:18px;">
				<div id="animals_level_2" style="display:none;background-color:#CFECEC;padding:1px 5px 1px 2px;margin-left:40px;font-size:0.9em;">
					<input type="radio" id="mammals" name="group" value="♞" <% $group eq "♞" ? "checked=\"checked\"" : ""%> /><label for="mammals"> Mammal</label>
					<input type="radio" id="birds" name="group" value="b" <% $group eq "b" ? "checked=\"checked\"" : ""%> /><label for="birds"> Bird</label>
					<input type="radio" id="fish" name="group" value="f" <% $group eq "f" ? "checked=\"checked\"" : ""%> /><label for="fish"> Fish</label>
					<input type="radio" id="insects" name="group" value="i" <% $group eq "i" ? "checked=\"checked\"" : ""%> /><label for="insects"> Insect</label>
					<input type="radio" id="nematodes" name="group" value="w" <% $group eq "w" ? "checked=\"checked\"" : ""%> /><label for="nematodes"> Nematode</label>
				</div>
				<img style="display:none;" src="/images/question-12.png" id="animals-matrix" />
				<div id="plants_level_2" style="display:none;background-color:#CFECEC;padding:1px 5px 1px 2px;margin-left:40px;font-size:0.9em;">
					<input type="radio" id="monocotyledons" name="group" value="m" <% $group eq "m" ? "checked=\"checked\"" : ""%> /><label for="monocotyledons"> Monocotyledon</label>
					<input type="radio" id="dicotyledons" name="group" value="d" <% $group eq "d" ? "checked=\"checked\"" : ""%> /><label for="dicotyledons"> Dicotyledon</label>
				</div>
				<img style="display:none;" src="/images/question-12.png" id="plant-matrix" />
			</div>
			<div style="clear:both;height:2px;">&nbsp;</div>
		 </div>
	
		<div class="conNewPro_title">Select Sequence Source *</div>

		<div class="conNewPro_label">
			<input class="conRadiobox_align" type="radio" name="seq_src" id="seq_src_upload" value="upload" onclick="select_source(this)" <% $seq_src eq "upload" ? "checked=\"checked\"" : ""%>/><label for="seq_src_upload">
			Upload a sequence file in <a href="http://en.wikipedia.org/wiki/FASTA_format#Format" target="_blank">FASTA format</a> <small>(max <% $seq_max_len %>kb)</small></label>:
			<input class="conStylized_box1" type="file" id="seq_file" name="seq_file"  onchange="set_source('upload')" onclick="set_source('upload')"  />
		</div>

		<div class="conNewPro_label">
			<input class="conRadiobox_align" type="radio" name="seq_src" id="seq_src_paste" value="paste" onclick="select_source(this)" <% $seq_src eq "paste" ? "checked=\"checked\"" : ""%> /><label for="seq_src_paste">
			Enter a sequence in <a href="http://en.wikipedia.org/wiki/FASTA_format#Format"  target="_blank">FASTA format</a> <small>(max <%$seq_max_len%>kb)</small></label>:
			<textarea class="conStylized_box2" name="sequence" id="notebox" rows="4" cols="30" onfocus="set_source('paste');"><% $sequence |html %></textarea>
		</div>

		<div class="conNewPro_label">
			<input class="conRadiobox_align" type="radio" name="seq_src" id="seq_src_sample" value="sample" onclick="select_source(this)" <% $seq_src eq "sample" ? "checked=\"checked\"" : ""%>/>
			<label for="seq_src_sample">Select a sample sequence:</label>
			<select class="conStylized_box3" id="specie" name="species" size="5" style="width: 370px;" onchange="set_source('sample');">
% for my $ss (@samples) {
%	my $sel = $seq_src eq "sample" && $ss->id == $species ? "selected=\"selected\"" : "";
			<option id="o<%$ss->id%>" value="<% $ss->id %>" extra="" clade="<% $ss->{clade} ? $ss->{clade} : "o" %>"\
			<% $sel %>><% $ss->organism %> (<% $ss->common_name %>) <% $ss->segment %>, <% nicebasepairs($ss->sequence_length) %></option>
% }
			</select>
		</div>
		
		<div class="conNewPro_title" style="padding-top:10px;">Add Evidence</div>
		<div class="conNewPro_label">
			<input class="conRadiobox_align" type="radio" name="add_evidence" id="add_evidence" onclick="$('gff_container').toggle();"/>
			<label for="add_evidence" >Import GFF from Green Line or Maker</label>
		</div>

		<div id="con_star" style="font-size:0.8em;margin-top:10px;font-style:italic"><b>*</b> Denotes required information</div>

        
	</div>
	  <!--END of ID conProject_newLeft-->
      </div><!--END of ID conRedlineLeft1-->

<div id="conRedlineRight1">

 <div id="conProject_newRight"> 
            	<div class="conNewPro_title">Name Your Project *</div>   
				<div class="conNewPro_label1">
					<label>Project title:</label>
					<input class="conStylized_box4" type="text" name="name" id="name" value="<% $name|html%>" maxlength="40" />
 				</div>
				<div style="clear: both;">&nbsp;</div>
                 <div class="conNewPro_title">Organism *</div>   
            		 <div class="conNewPro_label1">
						<label>Scientific name (genus species):</label>
								<input class="conStylized_box4" type="text" name="organism" id="organism" value="<% $organism |html%>" maxlength="128"  />
 				 </div>  
                 
            		 <div class="conNewPro_label1">
						<label>Common name:</label>
								<input class="conStylized_box4" type="text"name="common_name" id="common_name" value="<% $common_name |html%>" maxlength="128"   />
 				 	</div> 
            <!-- class: monocot, dicot, other, unkown was here-->
		<div style="clear: both;">&nbsp;</div>
        <div class="conNewPro_title">Description</div>
        <div class="conNewPro_label">Total characters (max.140): &nbsp;<span id="desc_len"><% length $description %></span></div>
		<div>
			<textarea id="description" name="description" cols="40" class="conStylized_box5" onkeyup="check_description_length(event);"
				style="width:300px; margin-left: 0px"><% $description |html %></textarea>
		</div>		
				<div style="clear: both;">&nbsp;</div> <div style="clear: both;">&nbsp;</div> 
              <div class="conBT_continue"><a id="step_one_btn" href="javascript:void(0);" onclick="javascript:step_one();"></a></div>
              <div style="clear: both;">&nbsp;</div>

            </div><!--END of ID conProject_newRight-->
			

<!-- Start Popup for Adding Evidence -->
<!-- Need to add variables / handling for $gff_evid -->
<div id="gff_container" style="position:absolute;left:-310px;bottom:75px;display:none">
	<div id="gff_container_inner">
		<div id="top">
			<span style="margin-right:12px;">Import from Green Line:</span>
			<input id="import_from_gl_browse" type="button" value="Browse" onclick="show_gl_data();"/>
			<input id="import_from_gl_cancel" type="button" value="Cancel" onclick="close_gl_data();" style="display:none;"/>
		</div>	
		<div id="evidence_uploads">
			<div>Import GFF data:                     
				<input class="conStylized_box1" type="file" id="evid_gff" name="evid_gff"/>
			</div>
			<div style="clear:both">&nbsp</div>
			<div style="text-align:center;">
				<input type="button" value="  Cancel  " class="input" onclick="$('gff_container').hide();" /> 
				<input type="submit" value="  Upload  " class="input" style="margin-left:10px;" />
			</div>
		</div>
		<div id="import_from_gl_space" style="display:none;"></div>
	</div>
</div>
<!-- End Popup for Adding Evidence -->
			
			
			
			
</div><!--END of ID conRedlineRight1-->
% if ($s->{new_user}) {
<input type="hidden" id="isnew" value="1" />
% }
</form>
<input type="hidden" id="test2" value="1" />
</div><!--END of ID container_rightContent-->

<script type="text/javascript">
	Event.observe(window, Prototype.Browser.IE ? 'load' : 'dom:loaded', function() {
		new Tip('animals-matrix', "<table id='matrix-table'><tr><th class='label'>&nbsp;</th><th>Augustus</th><th>FGenesH</th><th>SNAP</th></tr><tr><td class='label'>Mammal</td><td>H. sapiens</td><td>H. sapiens</td><td>Mammal</td></tr><tr><td class='label'>Fish</td><td>C. milii</td><td>F. rubripes</td><td>n/a</td></tr><tr><td class='label'>Insect</td><td>D. melanogaster</td><td>D. melanogaster</td><td>D. melanogaster</td></tr><tr><td class='label'>Bird</td><td>n/a</td> <td>G. Gallus</td><td>n/a</td></tr><tr><td class='label'>Nematode</td><td>C. elegans</td><td>C. elegans</td><td>C. elegans</td></tr></table><p style='padding:10px 15px;'>Gene predictors are optimized by using example data. This table shows the species used to train gene prediction algorithms for each class of organism. Training datasets are sparse as the number of fully sequenced genomes are limited. Choosing a category that is most closely related to your genome may yield better results. You may also rerun your data with alternative choices for comparison.</p>", {
			className: 'blue_help_wide',
			hook: {target:'bottomRight', tip:'topLeft'},
		});
		new Tip('plant-matrix', "<table id='matrix-table' style='width:407px;'><tr><th class='label'>&nbsp;</th><th>Augustus</th><th>FGenesH</th><th>SNAP</th></tr><tr><td class='label'>Monocotyledon</td><td>Z. mays</td><td>Monocots</td><td>O. sativa</td></tr><tr><td class='label'>Dicotyledon</td><td>A. thaliana</td><td>A. thaliana</td><td>A. thaliana</td></tr> </table><p style='padding:10px 15px;'>Gene predictors are optimized by using example data. This table shows the species used to train gene prediction algorithms for each class of organism. Training datasets are sparse as the number of fully sequenced genomes are limited. Choosing a category that is most closely related to your genome may yield better results. You may also rerun your data with alternative choices for comparison.</p>", {
			className: 'blue_help_wide',
			hook: {target:'bottomRight', tip:'topLeft'},
		});
	});
	
</script>

%#-----------------------------------------------------------------
<%args>
	$name => ''
	$seq_src => ''
	$species => ''
	$organism => ''
	$common_name => ''
	$group => ''
	$sequence => ''
	$description => ''
	$otype => ''
</%args>
<%once>
	use DNALC::Pipeline::Utils qw/nicebasepairs random_string/;
	use DNALC::Pipeline::Config ();
	use DNALC::Pipeline::Sample ();
	use DNALC::Pipeline::App::Utils ();
	use DNALC::Pipeline::App::ProjectManager ();
	use DNALC::Pipeline::App::WorkflowManager ();
	use Carp;
	use IO::File ();
	use Data::Dumper;
	
	my %clade_map = (
		animals => [qw/♞ f i w/],
		fungi => [qw/x/],
		plants => [qw/d m/],
	);
</%once>

<%init>
$m->session->{pipeline} ||= {};
my $s = $m->session->{pipeline};

my $cf = DNALC::Pipeline::Config->new;
my $pcf = $cf->cf('PIPELINE');
my $seq_max_len = int(($pcf->{sequence_length} || 50_000) / 1_000);

#my @samples = DNALC::Pipeline::Sample->search( active => 1, {order_by => "organism, segment, start"});
my @samples = ($otype && defined $clade_map{$otype} && $group)
				? DNALC::Pipeline::Sample->get_by_type( $group )
				: (); 

my @err = ();

if ($r->method eq 'POST') {

	my $file2process;
	my $converted_to_dna = 0;
	my $sequence_id = "";

	for ($name, $organism, $common_name) {
		$_ =~ s/\s+/ /g;
		$_ =~ s/^\s+//;
		$_ =~ s/\s+$//;
	}

	$organism = ucfirst lc $organism;

	unless ($name) {
		push @err, "Project title is missing";
	}

	unless ($organism && $common_name) {
		push @err, "Organism name is missing or incorrectly formatted!";
	}
	elsif($organism =~ /[^a-z0-9\s]/i || $common_name =~ /[^a-z0-9\s-]/i) {
		push @err, "Organism name may contain only letters, digits and spaces";
	}

	if ($seq_src eq "sample" && !defined $sequence) {
		push @err, "Sequence is missing", $/;
	}

	unless (@err) {
		$common_name = lc $common_name;

		if ($seq_src eq "sample") {
			my $sample = DNALC::Pipeline::Sample->new($species);
			if (!$sample) {
				push @err, "Sample organism wasn't found!";
			} 
			else {
				$file2process = $sample->sample_dir . '/fasta.fa';
			}
		}
		elsif ($seq_src eq "paste") {
		
			my $seq = $sequence;
			if (my @ids = ($seq =~ m/^((?:>|;).*)/mg)) {
				$sequence_id = join "\n", @ids;
			}
			$seq =~ s/^(?:>|;).*//mg;
			$seq =~ s/(?:\d|\s)+//g;

			if ($seq =~ /([^actugn]+)/i) {
				push @err, "Sequence contains invalid chars: [". uc ($1) . "]";
			}
			else {
			
				#make it DNA
				if ($seq =~ tr/uU/tT/) {
					$converted_to_dna = 1;
				}

				# make it FASTA
				$seq = "> fasta\n" . $seq;

				$file2process = $pcf->{upload_dir} . '/' . random_string(7,7);
				my $out = IO::File->new;
				if ($out->open($file2process, "w")) {
					print $out $seq;
					undef $out;
				}
			}
		}
		elsif ($seq_src eq "upload") {

			my $st = DNALC::Pipeline::App::Utils->save_upload( {
						r => $r, 
						param_name => 'seq_file',
						clean_sequence => 1,
					});

			if ($st->{status} eq "fail") {
				push @err, "Unable to upload file: ". $st->{message};
			}
			else {
				$file2process = $st->{path};
				$converted_to_dna = $st->{converted_to_dna};
				$sequence_id = $st->{sequence_id};
			}
		}

		my @subnames = split /\s/, $organism;
		unless (2 == @subnames) {
			push @err, "Organism name should contain only two words: genus and species";
		}

		if (length($description) > 140) {
			push @err, "Description should be shorter or equal to 140 characters"
		}
	}

	my ($pm, $proj);
	unless (@err) {
		$pm = DNALC::Pipeline::App::ProjectManager->new;
		my ($proj) = $pm->search(user_id => $s->{user_id}, name => $name);
		if ($proj) {
			my $mp = $proj->master_project;
			unless ($mp->archived) {
				push @err, "You already have a project named [$name]";
			}
		}
		else {
			my $conflicts = $pm->get_organism_conflicts({
					organism => $organism, 
					common_name => $common_name, 
					user_id => $s->{user_id}
				});
			#print STDERR ">>> conflicts: ", Dumper( $conflicts ), $/;
			for (@$conflicts) {
				push @err, "You may use <a href=\"#\" onclick=\"javascript:use_organism(this);\">" 
						. "<em>$_->{organism}, $_->{common_name}</em></a> as organism.";
			}
		}
	}
	# create the project 
	unless (@err) {
		my $rc = DNALC::Pipeline::App::Utils->process_input_file( $file2process );
		#print STDERR Dumper( $rc ), $/;
		if ($rc->{status} eq "success") {
			
	 		my $st = $pm->create_project ({
						user_id => $s->{user_id},
						seq => $rc->{seq},
						name => $name,
						organism => $organism,
						common_name => $common_name,
						sample => $species,
						clade => $group,
						description => $description,
					});
			if ($st->{status} eq 'success') {
				print STDERR  "New PID = ", $pm->project, $/;
				$proj = $pm->project;
			}
		}
		else {
			push @err, "Unable to process sequence data!";
			$proj = undef;
		}

		#do some cleaning.. rm any temporary file.. uploaded/pasted..
		if ($seq_src ne 'sample') { 
			# remove temporary file
			print STDERR  "Removing file: ", $file2process, $/;
			unlink $file2process;
		}

		
		unless ($proj) {
			push @err, "Error creating the project.";
			#$wfm->set_status('upload_fasta', 'error');
		}
		else {
			my $wfm = DNALC::Pipeline::App::WorkflowManager->new( $proj );
			$wfm->set_status('upload_fasta', 'done');
			print STDERR "project_created: id = ", $proj, ",\tname = ", $proj->name, $/;

			if ( $sequence_id ) {
				$pm->log("Saved sequence's id: " . $sequence_id);
			}

			if ($converted_to_dna) {
				$s->{converted_to_dna} = $proj->id;
				$pm->log("RNA converted to DNA.");
			}

			# no longer a new user (it has projects now...)
			delete $s->{new_user} if exists $s->{new_user};

			# redirect user to console
			$m->redirect('./console.html?pid=' . $proj->id);

		} #end else
	} #end unless
} #end if POST

</%init>

%#-----------------------------------------------------------------
<%attr>
	js => ['create_project.js', 'prototip/prototip.js']
	css => ['prototip.css']
	load_window_ui => 1
	current_section => 'red'
</%attr>

%# vim: ft=mason
